polar-core/src/data_filtering.rs

```diff
@@ -0,0 +1,799 @@
+                joined_cycle.insert(l);
+                joined_cycle.insert(r);
+                continue 'cycles;
+            }
+        }
+        // Create new cycle if we couldn't
+        let mut new_cycle = HashSet::new();
+        new_cycle.insert(l);
+        new_cycle.insert(r);
+        joined_cycles.push(new_cycle);
+    }
+
+    let mut next_id = 0;
+    let mut get_id = move || {
+        let id = next_id;
+        next_id += 1;
+        format!("{}", id)
+    };
+
+    // Give each cycle an id
+    let mut variables: HashMap<String, HashSet<Symbol>> = HashMap::new();
+    for joined_cycle in joined_cycles {
+        let id = get_id();
+        variables.insert(id, joined_cycle);
+    }
+
+    // Substitute in relationships
+    let mut parent_ids = vec![];
+    'parent_relationships: for (parent, _, _) in &var_info.field_relationships {
+        for (id, set) in &mut variables {
+            if set.contains(parent) {
+                parent_ids.push(id.clone());
+                continue 'parent_relationships;
+            }
+        }
+        // Create a new set if we didn't find one.
+        let new_id = get_id();
+        let mut new_set = HashSet::new();
+        new_set.insert(parent.clone());
+        variables.insert(new_id.clone(), new_set);
+        parent_ids.push(new_id);
+    }
+    let mut child_ids = vec![];
+    'child_relationships: for (_, _, child) in &var_info.field_relationships {
+        for (id, set) in &mut variables {
+            if set.contains(child) {
+                child_ids.push(id.clone());
+                continue 'child_relationships;
+            }
+        }
+        // Create a new set if we didn't find one.
+        let new_id = get_id();
+        let mut new_set = HashSet::new();
+        new_set.insert(child.clone());
+        variables.insert(new_id.clone(), new_set);
+        child_ids.push(new_id);
+    }
+
+    // If a.b = c and a.b = d, that means c = d.
+    // @Sorry(steve): Wow, what a loop. Maybe just loop over indexes instead.
+    let mut new_unifies: Vec<(String, String)> = vec![];
+    for (i, ((parent_id1, child_id1), (_, field1, _))) in parent_ids
+        .iter()
+        .zip(child_ids.iter())
+        .zip(var_info.field_relationships.iter())
+        .enumerate()
+    {
+        for (j, ((parent_id2, child_id2), (_, field2, _))) in parent_ids
+            .iter()
+            .zip(child_ids.iter())
+            .zip(var_info.field_relationships.iter())
+            .enumerate()
+        {
+            if i != j && parent_id1 == parent_id2 && field1 == field2 && child_id1 != child_id2 {
+                // Unify children
+                new_unifies.push((child_id1.clone(), child_id2.clone()));
+            }
+        }
+    }
+
+    // @TODO(steve): There are absolutely bugs in here.
+    // If we're turning 0 into 1 and then 0 into 2 it'll just blow up
+    // not correctly turn 0 and 1 into 2. Needs some tests.
+    for (x, y) in &new_unifies {
+        if x != y {
```

Comment:
Why is this logic different than the original logic you used when building `variables`?

Reply:
Because we're dealing with ids now instead of variables. We're merging two different variable sets.